### ISSUE TITLE
New feature: Huber location estimator

### ISSUE THREAD

**Issue by TimotheeMathieu** [scipy/scipy#13826]
**Is your feature request related to a problem? Please describe.**
The problem is to have an estimator of location (i.e. the mean) in 1 dimension that is not disrupted by outliers data and that is efficient in heavy-tailed setting. This is partly solved by the median but there are estimators that are much more efficient than the median when, for example, we want to estimate the mean of a corrupted Gaussian distribution.

**Describe the solution you'd like**

I'd like to implement Huber's estimator of location. This is an M-estimator, it is already [included in R library "MASS"](https://www.rdocumentation.org/packages/MASS/versions/7.3-53.1/topics/huber) for instance. This is similar but different from [statsmodel implementation](https://www.statsmodels.org/stable/generated/statsmodels.robust.scale.Huber.html) in part because statsmodel implementation is more complex than what I propose and it suffers a low breakdown point (number of outliers tolerated by the estimator). 

Huber estimator is very famous in robust statistics and has been used for quite some time now. In particular, one can look at the seminal article [1] but also at the books [2], [3] and any google scholar search will tell you that it is widely used.

There are several ways to compute this estimator but one of the simplest is the iterative reweighting algorithm found in [2].

**Describe alternatives you've considered**
None

**Additional context (e.g. screenshots)**
None

[1] Robust Estimation of a Location Parameter,  Peter J. Huber, Ann. Math. Statist (1964)
[2] Robust Statistics, Peter J. Huber and Elvezio M. Ronchetti , Wiley Series in Probability and Statistics (2009)
[3] Robust Statistics: The Approach Based on Influence Functions, Frank R. Hampel, Elvezio M. Ronchetti, Peter J. Rousseeuw, Werner A. Stahel, Wiley (2011)

**Comment by tupui** [comment]
Interesting feature. I am not sure if we want to include it if there is already something in statsmodels though. We try to not duplicate things.

@josef-pkt thoughts?

**Comment by TimotheeMathieu** [comment]
Further arguments to sell this feature:
 
A few examples : 
```
>>> import numpy as np
>>> import statsmodels.api as sm
>>> data = np.hstack([np.random.normal(size=10), 10, 12, 30, 100, 200])
>>> sm.robust.scale.huber(data)
```
result in ValueError "joint estimation of location and scale failed to converge in 30 iterations"

and 

```
>>> data = np.hstack([np.random.normal(size=15),100, 200, 200, 200, 200, 200])
>>> sm.robust.scale.huber(data)
(array(52.42405847), array(98.9021249))
```

so this is very much not robust and this is due to the fact that they estimate scale location jointly. In contrast, in R we get
```
 > library(MASS)
> data = c(rnorm(10), 10, 12, 30, 100, 200)
> 
> huber(data)
$mu
[1] 1.45101

$s
[1] 1.954187

> data = c(rnorm(15),100, 200, 200, 200, 200, 200)
> 
> huber(data)
$mu
[1] 0.9972582

$s
[1] 0.7729739
```

And conceptually speaking, the use of this estimator is close to what we do with the median and it is more basic than linear models and such so I think it is in fact a better fit for scipy than for statsmodel.

**Comment by josef-pkt** [comment]
If you want to estimate the mean with statsmodels, then use the RLM class.
It implements robust M-estimators with several objective functions "norms" including Huber, with options for scale methods including mad.
Based on the docstring of MASS huber, it uses MAD scale and not huber scale. 
statsmodels huber implements "Huber's proposal 2" which AFAIR uses an M-estimator for scale that is less robust than MAD, but more efficient.

So I guess this would mostly be a duplication of statsmodels without the extra features like standard errors and other post estimation results.

Aside: AFAIK, scipy least_squares allows for Huber objective function, and there is an example somewhere.

**Comment by tupui** [comment]
Thanks for the information! I will close the issue then.

@TimotheeMathieu if you feel this should really be included in SciPy, feel free to re-open by adding more justification and motivation since this exists in statsmodels. Or maybe if this is somehow in SciPy, I see `scipy.special.huber` but I did not check whether this related or not.

**Comment by TimotheeMathieu** [comment]
Ok, I reopen just to give you some of my experiments and to explicit the code for someone searching to compute Huber's estimator in python, you are right that least squares does the job even though least-squares is a slow because it is not tailored for this specific problem.

`scipy.special.huber` is huber loss function, this is linked but there is an optimization algorithm used to compute huber estimator.

Here are some experiments, the result is that because the algorithms are not ad-hoc for this task, they are rather slow. 
Also I think that these way to compute huber estimator are not as easily accessible as  they should be, I do robust statistics for quite some time and I never noticed that we could do it this way, and a user not so familiar with scipy would not know that there is a parameter in least-squares that allow for a mean computation with huber weights.

```
>data = np.hstack([2+np.random.normal(size=15),100, 200, 200, 200, 200, 200])
>from scipy.optimize import least_squares
>def huber_scipy(data):
    scaling = np.median(np.abs(data-np.median(data)))
    res = least_squares(lambda x : (data-x)/scaling, np.median(data), loss="huber")
    return res.x
>huber_scipy(data)
array([2.65093165])

```
This works. 

```
>import statsmodels.api as sm
>def huber_statsmodel(data):
    rlm_model = sm.RLM(data, np.ones(len(data)), M=sm.robust.norms.HuberT())
    rlm_results = rlm_model.fit()
    return rlm_results.params
>huber_statsmodel(data)
array([3.40811741])
```
this is not arbitrarily large but (I don't know why) this is not as good as using scipy least squares. Now, for the computation time :

 ```
>data= np.hstack([2+np.random.normal(size=10000),100, 200, 200, 200, 200, 200])
>init_time = time.time()
>huber_scipy(data)
>print(time.time()-init_time)
>init_time = time.time()
>huber_statsmodel(data)
>print(time.time()-init_time)
>init_time = time.time()
>huber.estimate(data)
>print(time.time()-init_time)
0.011896371841430664
0.04393124580383301
0.0015838146209716797
```
where the last huber.estimate is my method that estimates `2.64`, this is similar to scipy's estimate. 
As a summary you are right, this could be seen as a duplicate of least-squares, but my method is faster and it is very important to have a fast algorithm in my opinion when trying to estimate the mean because this can be used with very large datasets and a very large number of times.

**Comment by tupui** [comment]
Thanks for the information.

Then I would suggest trying to have a look first at why the current version is slow compared to your version (`least_squares` is doing a lot of checks, I suspect this is the main difference). Can you elaborate on why it's important that this estimator is fast? For me to understand in which context this can be useful. Because if this is really critical, we can always go to Cython or Pythran.

Since the functionality is available, I agree that it could be advertised. At least in the documentation with something along the line of your example. Does it sounds good @josef-pkt?

**Comment by TimotheeMathieu** [comment]
Thank you for reopening it for me. 

My implementation is in cython, this may be a reason for its speed, but I'm new to cython so maybe there is still room to improve my implementation.

The speed is not always critical but it can be, the idea is that Huber estimator is an estimator of the mean that is efficient even in a corrupted and Heavy-tailed setting. Then, for a robust application we may want to replace all the calls to the empirical mean by calls to Huber estimator. 
Now, think about how many calls to the empirical mean you make when training a Machine Learning algorithm for instance, this is a lot. For instance, Huber estimator can be used to do robust KMeans (ref [1]) and such applications and in this context, with large dataset, you want Huber estimator to be fast.

So even though this is not critical in all applications, Big data applications are an important example that I know in which speed is an important factor. What I don't understand is that statsmodel method is based on the same principle as my algorithm and should perform similarly but I didn't succeed in making it work.

[1] Empirical risk minimization for heavy-tailed losses,  Christian Brownlees, Emilien Joly, Gábor Lugosi in Ann. Stats. (2015)

**Comment by josef-pkt** [comment]
Estimating just mean (and required scale) can use much simpler code, than generic multiparameter/regression versions.

For fixed scale, the mean estimate is just a simple iteration to reach the minimum, or alternating scale and mean iteration to get both. 

(after next message)

I don't think cython is worth it. 
This is just computing a weighted mean a few times in the iteration, which is mostly numpy.

**Comment by TimotheeMathieu** [comment]
Well I had a 10x speed up using cython but maybe my python implementation was not good ? 
My hypothesis was that cython was good for score function computation (score function = derivative of loss function) which is needed to compute the weights in the weighted mean.

@josef-pkt : You are right that the rlm algo is overkill for this problem but I am still curious of how you would use RLM to do a good estimation. As you can see, the rapid implementation I described previously didn't work so well compared to least-squares but I don't now why, do you ?

**Comment by josef-pkt** [comment]
> As you can see, the rapid implementation I described previously didn't work so well compared to least-squares but I don't now why, do you ?

My guess is the difference comes from the scale estimation method. In the leastsquares version you use MAD but never update inside mean optimization.
statsmodels uses mad as starting value but by default updates scale in each iteration. There is an `update_scale=False`, but I haven't tried that.
(Default scale_est='mad', I thought it was Huber-scale. I haven't looked at those parts in a while)

If you really want higher breakdown methods, than it is better to use a redescending `norm`, which puts zero weight on large outliers.

**Comment by TimotheeMathieu** [comment]
You are right, `update_scale=False` does the trick, I get the same result then. 

Redescending estimators have their own set of pro/cons and Huber estimator with fixed scale has 0.5 breakdown point which is the same that any redescending M-estimator.

**Comment by TimotheeMathieu** [comment]
I made a colab notebook so you can see for yourself and test the implementation.

https://colab.research.google.com/drive/1oXcVgRambZaPydOkj0vUBKSQyJ1tSmWH?usp=sharing

**Comment by josef-pkt** [comment]
>  `mu = np.average(X, axis=0, weights=w/np.sum(w))`

dividing by `np.sum(w)` is redundant according to the `average` docstring

I would have used `w.dot(X) / w.sum()`  which might be a bit faster.

**Comment by josef-pkt** [comment]
your `class m_estimator` doesn't use any scale AFAICS ?

**Comment by TimotheeMathieu** [comment]
> I would have used w.dot(X) / w.sum() which might be a bit faster.
I did this and this is not much faster. But still it is a bit faster, thanks. 

> your class m_estimator doesn't use any scale AFAICS ?
Yes my `class m_estimator` doesn't use any scale. We can put in a `mad` scale as it is often done and it would not impact the speed of convergence by a lot, or we can tune the constant (for instance in a Machine Learning Framework we may want to tune with cross validation). The constant scale is often set to 1.5 or something similar to be efficient on a standard normal (or corrupted standard normal) distribution. 

One may also want to tune the constant scale by hand as this scale is also a robustness parameter that tells you how much influence outliers have and what is considered an outlier (small scale parameter would get you small influence for outliers).

**Comment by TimotheeMathieu** [comment]
Ok, as this seems to be already in part implemented in scipy and statsmodel (up to speed of computation), I am closing.
